apps/emqx_bridge_kafka/src/emqx_bridge_kafka_impl_producer.erl

```diff
@@ -474,3 +513,108 @@ maybe_install_wolff_telemetry_handlers(ResourceID) ->
+    %% the value is already validated by schema, so we do not validate it again.
+    emqx_placeholder:preproc_tmpl(HeadersTmpl).
+
+preproc_ext_headers(Headers) ->
+    [
+        {emqx_placeholder:preproc_tmpl(K), preproc_ext_headers_value(V)}
+     || #{kafka_ext_header_key := K, kafka_ext_header_value := V} <- Headers
+    ].
+
+preproc_ext_headers_value(ValTmpl) ->
+    %% the value is already validated by schema, so we do not validate it again.
+    emqx_placeholder:preproc_tmpl(ValTmpl).
+
+proc_ext_headers(ExtHeaders, Msg) ->
+    lists:filtermap(
+        fun({KTks, VTks}) ->
+            try
+                Key = proc_ext_headers_key(KTks, Msg),
+                Val = proc_ext_headers_value(VTks, Msg),
+                {true, {Key, Val}}
+            catch
+                throw:placeholder_not_found -> false
+            end
+        end,
+        ExtHeaders
+    ).
+
+proc_ext_headers_key(KeyTks, Msg) ->
+    RawList = emqx_placeholder:proc_tmpl(KeyTks, Msg, #{return => rawlist}),
+    list_to_binary(
+        lists:map(
+            fun
+                (undefined) -> throw(placeholder_not_found);
+                (Key) -> bin(Key)
+            end,
+            RawList
+        )
+    ).
+
+proc_ext_headers_value(ValTks, Msg) ->
+    case emqx_placeholder:proc_tmpl(ValTks, Msg, #{return => rawlist}) of
+        [undefined] -> throw(placeholder_not_found);
+        [Value] -> Value
+    end.
+
+kvlist_headers([], Acc) ->
+    lists:reverse(Acc);
+kvlist_headers([#{key := K, value := V} | Headers], Acc) ->
+    kvlist_headers(Headers, [{K, V} | Acc]);
+kvlist_headers([#{<<"key">> := K, <<"value">> := V} | Headers], Acc) ->
+    kvlist_headers(Headers, [{K, V} | Acc]);
+kvlist_headers([{K, V} | Headers], Acc) ->
+    kvlist_headers(Headers, [{K, V} | Acc]);
+kvlist_headers([BadHeader | _], _) ->
+    throw({bad_kafka_header, BadHeader}).
+
+-define(IS_STR_KEY(K), (is_list(K) orelse is_atom(K) orelse is_binary(K))).
+formalize_kafka_headers(Headers, none) ->
+    %% Note that we will drop all the non-binary values in the NONE mode
+    [{bin(K), V} || {K, V} <- Headers, is_binary(V) andalso ?IS_STR_KEY(K)];
+formalize_kafka_headers(Headers, json) ->
+    lists:filtermap(
+        fun({K, V}) ->
+            try
+                {true, {bin(K), emqx_utils_json:encode(V)}}
+            catch
+                _:_ -> false
+            end
+        end,
+        Headers
+    ).
+
+merge_kafka_headers(HeadersTks, ExtHeaders, Msg) ->
+    case emqx_placeholder:proc_tmpl(HeadersTks, Msg, #{return => rawlist}) of
+        % Headers by map object.
+        [Map] when is_map(Map) ->
+            maps:to_list(Map) ++ ExtHeaders;
+        [KVList] when is_list(KVList) ->
+            kvlist_headers(KVList, []) ++ ExtHeaders;
+        %% the placeholder cannot be found in Msg
+        [undefined] ->
+            ExtHeaders;
+        [MaybeJson] when is_binary(MaybeJson) ->
+            try emqx_utils_json:decode(MaybeJson) of
```

Comment:
Nit: there's also `emqx_utils_json:safe_decode` that avoids the need for `try/catch`.